### DISCUSSION TITLE
[feature request] full n-of-m authentication support for SmartCard HSM

### DISCUSSION THREAD

**Discussion by mjourdan** [OpenSC/OpenSC#3246] [Ideas]
Hi,

A typical use case for using a HSM is for building up a PKI and keeping the root CA offline and protected from copying. Requiring m-of-n authentication on that HSM allows for a shared control by a group of key custodians over the root CA usage. Each key custodian would have to enter a PIN to authenticate with their respective device.

When using Smartcard HSM, this can be achieved with  `SmartCard Shell` for HSM initialization (as documented [here](https://www.smartcard-hsm.com/docs/n-of-m_Authentication_HowTo_V1_2015-09-07.pdf)) and `PKI-as-a-service` for CA usage (as shown [there](https://www.smartcard-hsm.com/2022/05/09/prepare-trustcenter-token.html)).

Would it be technically realistic to implement m-of-n pubkey authentication in OpenSC so that one could sign intermediate certificates and revocation lists using openssl and the pkcs11 engine, without the need of `PKI-as-a-service`?

In addition, I understand the following features [might be missing](https://github.com/OpenSC/OpenSC/discussions/3206) in sc-hsm-tool in order to not depend on `SmartCard Shell`(please correct me if I'm wrong):
- initialize CA HSM with pubkeys as the only authentification method enabled (ie, PIN authentication should be disabled)
- initialize authentication HSM so that resetting PIN with SO-PIN would be disallowed

Are others interested in such features?

As we don't have the skills to contribute code, we would like to know if anyone who has would be interested for paid work?

**Comment by CardContact** [comment]
We can do that as part of a customer project, but it is difficult to get that integrated in OpenSC.

The topic was discussed in #594, but there was virtually no demand from users.

A customer of us did integrate n-of-m PKA in EJBCA, but that is not available as open source. We are using PKA in the PKI-as-a-Service portal and have support for it in the JCE-Provider.

What we propose to do, is to add support in [s-hsm-embedded](https://github.com/CardContact/sc-hsm-embedded), as that already includes support for [RAMOverHTTP](https://www.openscdp.org/scriptingserver/remoteterminal.html).

The PKCS#11 module indicate CKF_PROTECTED_AUTHENTICATION_PATH, if an environment variable with an URL that points to an authentication service is defined. In that case, calling C_Login will not verify the PIN, but instead connect the token to the authentication service using RAMOverHTTP. Users can then use the authentication service to perform n-of-m PKA authentication and once complete, pass back control to the application that is using the PKCS#11 API.

The authentication service could be a trust center instance in the PKI-as-a-Service or the sc-hsm-management-server that we provide in the [CDN](https://www.smartcard-hsm.com/cdn.html) for that purpose.

**Reply by mjourdan** [reply]
> The topic was discussed in https://github.com/OpenSC/OpenSC/issues/594, but there was virtually no demand from users.

In the past 4 months, different persons seemed to have very similar use cases and expectations. There are two of us as seen on the Nitrokey forum, plus someone here who opened #3206. Surely all 3 of us would be interested in such features. Hopefully some of us would be interested enough to participate in funding the work or contributing by any other means?

@Viajaz, maybe you would like to jump in and share your thoughts?

> What we propose to do, is to add support in [s-hsm-embedded](https://github.com/CardContact/sc-hsm-embedded), as that already includes support for [RAMOverHTTP](https://www.openscdp.org/scriptingserver/remoteterminal.html).

Thanks for the opening. My understanding is that this approach would not help much as using the HSM for a simple root CA would still depend on the same external dependencies.

**Reply by Viajaz** [reply]
I haven't been able to get the feature working with any software as of yet but I haven't finished testing, our test project is on standby though. My experience with it thus far has not been positive.
